lib/route/link/bridge.c

```diff
@@ -1376,6 +1625,186 @@ struct rtnl_link_bridge_vlan *rtnl_link_bridge_get_port_vlan(struct rtnl_link *l
+ * @arg link		Link object of type bridge
+ * @arg instance	MST instance number
+ *
+ * @return The state (BR_STATE_*) on success, or a negative error code
+ * @return -NLE_INVAL 		link is NULL
+ * @return -NLE_OPNOTSUP 	Link is not a bridge
+ * @return -NLE_OBJ_NOTFOUND	MST instance not found
+ */
+int rtnl_link_bridge_get_mst_port_state(struct rtnl_link *link,
+					uint16_t instance)
+{
+	struct bridge_data *bd;
+	struct mst_state_entry *entry;
+
+	if (link == NULL)
+		return -NLE_INVAL;
+
+	IS_BRIDGE_LINK_ASSERT(link);
+
+	bd = bridge_data(link);
+	if (bd == NULL)
+		return -NLE_OPNOTSUPP;
+
+	entry = find_mst_state_entry(bd, instance);
+
+	if (entry == NULL)
+		return -NLE_OBJ_NOTFOUND;
+
+	return entry->state;
+}
+
+/**
+ * Delete the Multiple Spanning Tree (MST) port state for a given MST instance
+ * @arg link		Link object of type bridge
+ * @arg instance	MST instance number
+ *
+ * @return 0 on success or a negative error code
+ * @return -NLE_INVAL 		link is NULL
+ * @return -NLE_OPNOTSUP 	Link is not a bridge
+ * @return -NLE_OBJ_NOTFOUND	MST instance not found
+ */
+int rtnl_link_bridge_del_mst_port_state(struct rtnl_link *link,
+					uint16_t instance)
+{
+	struct bridge_data *bd;
+	struct mst_state_entry *entry;
+
+	if (link == NULL)
+		return -NLE_INVAL;
+
+	IS_BRIDGE_LINK_ASSERT(link);
+
+	bd = bridge_data(link);
+	if (bd == NULL)
+		return -NLE_OPNOTSUPP;
+
+	entry = find_mst_state_entry(bd, instance);
+
+	if (entry == NULL)
+		return -NLE_OBJ_NOTFOUND;
+
+	mst_state_entry_del(entry);
+
+	if (nl_list_empty(&bd->mst_list))
+		bd->ce_mask &= ~BRIDGE_ATTR_MST;
+
+	return 0;
+}
+
+/**
+ * Delete all Multiple Spanning Tree (MST) port state information
+ * @arg link		Link object of type bridge
+ *
+ * @return 0 on success or a negative error code
+ * @return -NLE_INVAL 		link is NULL
+ * @return -NLE_OPNOTSUP 	Link is not a bridge
+ */
+int rtnl_link_bridge_clear_mst_port_state_info(struct rtnl_link *link)
+{
+	struct bridge_data *bd;
+
+	if (link == NULL)
+		return -NLE_INVAL;
+
+	IS_BRIDGE_LINK_ASSERT(link);
+
+	bd = bridge_data(link);
+	if (bd == NULL)
+		return -NLE_OPNOTSUPP;
+
+	mst_list_clear(&bd->mst_list);
+	bd->ce_mask &= ~BRIDGE_ATTR_MST;
+
+	return 0;
+}
+
+/**
+ * Iterate over all Multiple Spanning Tree (MST) port state entries
+ * @arg link		Link object of type bridge
+ * @arg cb		Callback function
+ * @arg arg		User provided data argument to pass to the callback
+ *			function
+ *
+ * The callback function is called for each MST entry. It is passed the MST
+ * instance ID, state (BR_STATE_*), and an optional user provided data argument.
+ * MST entries should not be added or removed by the callback function.
+ *
+ * @return 0 on success or a negative error code
+ * @return -NLE_INVAL 		link or cb is NULL
+ * @return -NLE_OPNOTSUP 	Link is not a bridge
+ */
+int rtnl_link_bridge_foreach_mst_entry(
+	struct rtnl_link *link,
+	void (*cb)(uint16_t instance, uint8_t state, void *arg), void *arg)
```

Comment:
hm. This seems really cumbersome API... and also error prone, because the callback must guarantee not to modify the object while iterating... but ok. No better idea.

Reply:
I could use a (thread local) static variable that acts as a lock that is set when this function is called and unset when the function is exited. The modification functions could then bail out if the variable is set. There is also `nl_list_for_each_entry_safe` which maybe I should be using instead.

Reply:
hm...

I general, I find the way we track the mst_state_entry via a linked list not great. It has more overhead than just allocating (and `realloc()`) an array. Since we cannot do efficient O(1) lookup by index, it also does not lean to an API where we could have
```C
int rtnl_link_bridge_get_mst_entry_at(struct rtnl_link *link, size_t index, uint16_t *out_msti, uint8_t *out_state);
```

I think I am fine with it as-is.

Reply:
There is no significance to the order of the MST entries so I don't think there is much use in exposing a way to get them by index. I chose to just use a list because that's what some other modules use, and it made implementation easier. However I do agree that other data structures could be better suited (such as a hashmap or self-balancing tree). If the list implementation is outgrown, the implementation could be switched to one of these without changing the external API, since the list implementation is not exposed to the user.

Reply:
> There is no significance to the order of the MST entries so I don't think there is much use in exposing a way to get them by index.

While the index has no particular meaning in most regards, it would have a meaning for accessing the elements.

Just like the callback is invoked in a certain order. That order is also not particularly meaningful, but such order still exists and is visible to the user of the API.

Anyway.